### PR TITLE
/webapp/boards: fix unit tests and run in ci

### PR DESCRIPTION
**Pull request by calebroseland**
#### Summary

TL;DR: Fixes `/webapp/boards` unit tests, and enables them in CI (without breaking `/webapp/channels` type-checking as earlier revisions did).

<details>
<summary>Details</summary>

The `focalboard/mattermost-plugin/webapp`-`focalboard/webapp` merge and subsequent migration to monorepo has caused some issues in execution of `webapp/boards` unit tests. This remedies that by upgrading to Jest 29 and related packages, and ironing out any breaking changes and other assorted issues in moving to monorepo. The fixing-without-upgrading path was explored and did not prove to be worth it.

Very limited changes were made to runtime source code to fix various unit tests and typings:
1. `kanban.tsx`, `kanbanColumnHeader.tsx`, etc. now use `useIntl` instead of `injectIntl` & `intl` prop-passing
2. `editable.tsx` now uses `KeyboardEvent.key` instead of the deprecated `KeyboardEvent.keyCode`
3. `StylesConfig` typing was corrected for a few react-select instances
4. Env typing: upgraded to `react@17`, `react-dom@17`; downgraded to `@types/node@16`.

The remaining changes are all to test files, configs, and dev-deps.

[`jest-fail-on-console`](https://www.npmjs.com/package/jest-fail-on-console) was added to help surface silent/hidden errors—`undefined` Ids , `key`s missing in loops, async/cleanup errors, etc. `react-select-event`, a react-select `userEvent` equivalent is added but presently unused.

3 tests are skipped due to complexity in resolving testing env changes. Will be addressed after this PR.

Jest configuration moves out of package.json to its own `jest.config.js`.

Jest scripts add an explicit timezone (`Etc/UTC`) for time consistency; `--forceExit --detectOpenHandles --verbose` are  moved to the new `npm run test:debug` so main scripts can support multi-workers.

In the future we'll consider adding [`eslint-plugin-testing-library`](https://github.com/testing-library/eslint-plugin-testing-library) which will assist in modernizing and optimizing testing-library usage. Reports some 700 warnings currently, which was too much for this PR.

NPM `global-style` is enabled to isolate workspace dependencies more predictably.

</details>

#### Ticket Link
None

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: webapp/.npmrc [hunk 1]
```diff
@@ -1,2 +1,3 @@
 save-exact=true
 legacy-peer-deps=true
+global-style=true
```
Comment (calebroseland) [inline webapp/.npmrc:3]: This is the main change. There may be some dedupe cost, but only this or `--legacy-bundling` (which disables all deduping) paired with reinstalling `@types/react-select@3.0.19` in the channels workspace seemed to fix the latest cross-workspace module resolution issues.

File: webapp/package-lock.json [hunk 2]
```diff
@@ -4048,6 +3958,17 @@
...
+      "version": "3.0.19",
+      "resolved": "https://registry.npmjs.org/@types/react-select/-/react-select-3.0.19.tgz",
+      "integrity": "sha512-d+6qtfFXZeIOAABlVL1e50RZn8ctOABE4tFDxM6KW4lKuXgTTgLVrSik5AX9XjBjV7N80FtS6GTN/WeoXL9Jww==",
+      "dev": true,
+      "dependencies": {
+        "@types/react": "*",
+        "@types/react-dom": "*",
+        "@types/react-transition-group": "*"
+      }
+    },
```
Comment (calebroseland) [inline webapp/package-lock.json:3971]: Reinstalling `@types/react-select@3.0.19` after `global-style` was the latest whack-a-mole which seems to stabilize it (types and unit tests are passing with all workspaces together)

File: webapp/.npmrc (hunk 1 above)
Comment (hmhealey) [reply webapp/.npmrc:3]: Thanks for digging into this. I think this would've helped fix some other issues I've fought with as well. From what I can find, npm used to have this behaviour years ago then stopped doing it despite knowing there were were issues with it, and later re-added this as an option. I haven't found any docs suggesting this is at all a bad idea, and it works similarly in pnpm as well apparently.

That setting got replaced by `install-strategy=shallow` [in npm 9](https://github.blog/changelog/2022-10-24-npm-v9-0-0-released/) which we'll need to remember to change, but this sounds good to me.

**Review by hmhealey — APPROVED** [review]
Looks great! I'm hoping this way of installing dependencies helps avoid those issues that have popped up a few times already. I have one question about the Jest config, but it's not blocking even if we do want to change it

**Comment by calebroseland** [comment]
> I also had to reinstall @types/react-select@3.0.19. Then all tests ran successfully.

Did you run `npm run ...` directly? The make targets should do an `npm install` if needed.

File: webapp/boards/jest.config.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        "\\.(scss|css)$": "<rootDir>/__mocks__/styleMock.js",
+        "^bundle-loader\\?lazy\\!(.*)$": "$1",
+        "^src(.*)$": "<rootDir>/src$1",
+        "^i18n(.*)$": "<rootDir>/i18n$1",
+        "^static(.*)$": "<rootDir>/static$1",
+        "^react-day-picker(.*)$": "<rootDir>/node_modules/react-day-picker$1",
+        "^moment(.*)$": "<rootDir>/../node_modules/moment$1",
+        "^react$": "<rootDir>/../node_modules/react",
+        // "^react-redux$": "<rootDir>/../node_modules/react-redux",
+        // "^react-intl$": "<rootDir>/../node_modules/react-intl",
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply webapp/boards/jest.config.js:50]: Not now with the other module issues resolved, but they were needed before. Will remove.

**Comment by calebroseland** [comment]
@lindy65 This is largely just around tooling changes, but there are a few small runtime changes described in the desc. I'll leave it up to your judgement if QA-Deferred would be a better option or not.

**Review by lindy65 — APPROVED** [review]
Thanks @calebroseland - I've added the QA Deferred label 👍
